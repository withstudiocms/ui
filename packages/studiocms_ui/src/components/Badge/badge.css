.sui-badge {
  display: flex;
  flex-direction: row;
  gap: .5rem;
  align-items: center;
  padding: .125rem .75rem;
  border-radius: var(--radius-full);
  font-size: .875rem;
}

.sui-badge.semi {
  border-radius: var(--radius-md);
}

.sui-badge.primary {
  background-color: hsl(var(--primary-base));
  color: hsl(var(--text-inverted));
}

.sui-badge.primary.flat, .sui-badge.primary.outlined {
  background-color: hsla(var(--primary-flat));
  color: hsl(var(--primary-base));
}

.sui-badge.primary.outlined {
  border: 1px solid hsl(var(--primary-base));
}

.sui-badge.success {
  background-color: hsl(var(--success-base));
  color: hsl(var(--text-inverted));
}

.sui-badge.success.flat, .sui-badge.success.outlined {
  background-color: hsla(var(--success-flat));
  color: hsl(var(--success-base));
}

.sui-badge.success.outlined {
  border: 1px solid hsl(var(--success-base));
}

.sui-badge.warning {
  background-color: hsl(var(--warning-base));
  color: hsl(var(--text-inverted));
}

.sui-badge.warning.flat, .sui-badge.warning.outlined {
  background-color: hsla(var(--warning-flat));
  color: hsl(var(--warning-base));
}

.sui-badge.warning.outlined {
  border: 1px solid hsl(var(--warning-base));
}

.sui-badge.danger {
  background-color: hsl(var(--danger-base));
  color: hsl(var(--text-normal));
}

.sui-badge.danger.flat, .sui-badge.danger.outlined {
  background-color: hsla(var(--danger-flat));
  color: hsl(var(--danger-base));
}

.sui-badge.danger.outlined {
  border: 1px solid hsl(var(--danger-base));
}

.sui-badge.info {
  background-color: hsl(var(--info-base));
  color: hsl(var(--text-normal));
}

.sui-badge.info.flat, .sui-badge.info.outlined {
  background-color: hsla(var(--info-flat));
  color: hsl(var(--info-base));
}

.sui-badge.info.outlined {
  border: 1px solid hsl(var(--info-base));
}

.sui-badge.mono {
  background-color: hsl(var(--mono-base));
  color: hsl(var(--text-inverted));
}

.sui-badge.mono.flat, .sui-badge.mono.outlined {
  background-color: hsla(var(--mono-flat));
  color: hsl(var(--mono-base));
}

.sui-badge.mono.outlined {
  border: 1px solid hsl(var(--mono-base));
}

[data-theme="light"] .sui-badge.danger {
  color: hsl(var(--text-inverted));
}

[data-theme="light"] .sui-badge.info {
  color: hsl(var(--text-inverted));
}

.sui-badge.sm {
  padding: .125rem .75rem;
}

.sui-badge.sm {
  font-size: .75rem;
}

.sui-badge.sm > svg {
  width: 14px;
  height: 14px;
}

.sui-badge.lg {
  font-size: 1rem;
  padding: .25rem 1rem;
}

.sui-badge.lg {
  font-size: 1rem;
}

.sui-badge.lg > svg {
  width: 20px;
  height: 20px;
}
