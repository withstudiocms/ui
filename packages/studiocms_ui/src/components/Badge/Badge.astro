---
import type { StudioCMSColorway } from '../../utils/colors.js';
import './badge.css';
import type { HTMLAttributes } from 'astro/types';
import Icon from '../Icon/Icon.astro';
import type { HeroIconName } from '../Icon/iconType.js';

interface Props extends Omit<HTMLAttributes<'span'>, 'color'> {
	/**
	 * The color of the badge. Defaults to `primary`.
	 */
	color?: Omit<StudioCMSColorway, 'default'>;
	/**
	 * The icon to display in the badge.
	 */
	icon?: HeroIconName;
	/**
	 * The size of the badge. Defaults to `md`.
	 */
	size?: 'sm' | 'md' | 'lg';
	/**
	 * The variant of the badge. Defaults to `default`.
	 */
	variant?: 'default' | 'flat' | 'outline';
	/**
	 * The rounding of the badge. Full is a half-circle, semi is a quarter-circle. Defaults to `full`.
	 */
	rounding?: 'full' | 'semi';
	/**
	 * The label of the badge.
	 */
	label: string;
}

const {
	color = 'primary',
	icon,
	size = 'md',
	variant = 'default',
	rounding = 'full',
	label,
	...props
} = Astro.props;
---

<span class="sui-badge" class:list={[color, size, variant, rounding]} {...props}>
  {icon && <Icon name={icon} width={16} height={16} />}
  {label}
</span>
