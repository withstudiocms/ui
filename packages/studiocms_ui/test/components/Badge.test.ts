import { describe, expect } from 'vitest';
import Badge from '../../src/components/Badge/Badge.astro';
import { test } from '../fixtures/vitest/AstroContainer';

describe('Badge Component', () => {
	test('renders Badge component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Badge, 'Badge', { props: { label: 'Test Badge' } });
		expect(result).toMatchSnapshot();
	});

	test('renders Badge component with different colors', async ({ renderComponent }) => {
		const colors = ['primary', 'success', 'warning', 'danger', 'info', 'mono'];

		for (const color of colors) {
			const result = await renderComponent(Badge, `Badge-${color}`, {
				props: { label: `Badge ${color}`, color },
			});
			expect(result).toMatchSnapshot();
		}
	});

	test('renders Badge component with different sizes', async ({ renderComponent }) => {
		const sizes = ['sm', 'md', 'lg'];

		for (const size of sizes) {
			const result = await renderComponent(Badge, `Badge-${size}`, {
				props: { label: `Badge ${size}`, size },
			});
			expect(result).toMatchSnapshot();
		}
	});

	test('renders Badge component with different variants', async ({ renderComponent }) => {
		const variants = ['default', 'flat', 'outline'];

		for (const variant of variants) {
			const result = await renderComponent(Badge, `Badge-${variant}`, {
				props: { label: `Badge ${variant}`, variant },
			});
			expect(result).toMatchSnapshot();
		}
	});

	test('renders Badge component with icon', async ({ renderComponent }) => {
		const positions = [undefined, 'left', 'right'];

		for (const position of positions) {
			const result = await renderComponent(Badge, `Badge-with-icon-${position || 'default'}`, {
				props: {
					label: `Badge with Icon ${position || 'default'}`,
					icon: 'heroicons:academic-cap',
					iconPosition: position,
				},
			});
			expect(result).toMatchSnapshot();
		}
	});

	test('renders Badge component with different rounding', async ({ renderComponent }) => {
		const roundings = [undefined, 'full', 'semi'];

		for (const rounding of roundings) {
			const result = await renderComponent(Badge, `Badge-rounding-${rounding || 'default'}`, {
				props: { label: `Badge ${rounding || 'default'}`, rounding },
			});
			expect(result).toMatchSnapshot();
		}
	});
});
