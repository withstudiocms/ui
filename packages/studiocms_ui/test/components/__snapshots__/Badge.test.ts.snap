// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge Component > renders Badge component correctly 1`] = `"<span class="sui-badge primary md outlined full">  Test Badge  </span>"`;

exports[`Badge Component > renders Badge component with different colors 1`] = `"<span class="sui-badge primary md outlined full">  Badge primary  </span>"`;

exports[`Badge Component > renders Badge component with different colors 2`] = `"<span class="sui-badge success md outlined full">  Badge success  </span>"`;

exports[`Badge Component > renders Badge component with different colors 3`] = `"<span class="sui-badge warning md outlined full">  Badge warning  </span>"`;

exports[`Badge Component > renders Badge component with different colors 4`] = `"<span class="sui-badge danger md outlined full">  Badge danger  </span>"`;

exports[`Badge Component > renders Badge component with different colors 5`] = `"<span class="sui-badge info md outlined full">  Badge info  </span>"`;

exports[`Badge Component > renders Badge component with different colors 6`] = `"<span class="sui-badge mono md outlined full">  Badge mono  </span>"`;

exports[`Badge Component > renders Badge component with different rounding 1`] = `"<span class="sui-badge primary md outlined full">  Badge default  </span>"`;

exports[`Badge Component > renders Badge component with different rounding 2`] = `"<span class="sui-badge primary md outlined full">  Badge full  </span>"`;

exports[`Badge Component > renders Badge component with different rounding 3`] = `"<span class="sui-badge primary md outlined semi">  Badge semi  </span>"`;

exports[`Badge Component > renders Badge component with different sizes 1`] = `"<span class="sui-badge primary sm outlined full">  Badge sm  </span>"`;

exports[`Badge Component > renders Badge component with different sizes 2`] = `"<span class="sui-badge primary md outlined full">  Badge md  </span>"`;

exports[`Badge Component > renders Badge component with different sizes 3`] = `"<span class="sui-badge primary lg outlined full">  Badge lg  </span>"`;

exports[`Badge Component > renders Badge component with different variants 1`] = `"<span class="sui-badge primary md default full">  Badge default  </span>"`;

exports[`Badge Component > renders Badge component with different variants 2`] = `"<span class="sui-badge primary md flat full">  Badge flat  </span>"`;

exports[`Badge Component > renders Badge component with different variants 3`] = `"<span class="sui-badge primary md outline full">  Badge outline  </span>"`;

exports[`Badge Component > renders Badge component with icon 1`] = `"<span class="sui-badge primary md outlined full"> <svg style="min-width: 16px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" height="16" width="16" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M4.26 10.147a60 60 0 0 0-.491 6.347A48.6 48.6 0 0 1 12 20.904a48.6 48.6 0 0 1 8.232-4.41a61 61 0 0 0-.491-6.347m-15.482 0a51 51 0 0 0-2.658-.813A60 60 0 0 1 12 3.493a60 60 0 0 1 10.399 5.84q-1.345.372-2.658.814m-15.482 0A51 51 0 0 1 12 13.489a50.7 50.7 0 0 1 7.74-3.342M6.75 15a.75.75 0 1 0 0-1.5a.75.75 0 0 0 0 1.5m0 0v-3.675A55 55 0 0 1 12 8.443m-7.007 11.55A5.98 5.98 0 0 0 6.75 15.75v-1.5"/></svg> Badge with Icon default  </span>"`;

exports[`Badge Component > renders Badge component with icon 2`] = `"<span class="sui-badge primary md outlined full"> <svg style="min-width: 16px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" height="16" width="16" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M4.26 10.147a60 60 0 0 0-.491 6.347A48.6 48.6 0 0 1 12 20.904a48.6 48.6 0 0 1 8.232-4.41a61 61 0 0 0-.491-6.347m-15.482 0a51 51 0 0 0-2.658-.813A60 60 0 0 1 12 3.493a60 60 0 0 1 10.399 5.84q-1.345.372-2.658.814m-15.482 0A51 51 0 0 1 12 13.489a50.7 50.7 0 0 1 7.74-3.342M6.75 15a.75.75 0 1 0 0-1.5a.75.75 0 0 0 0 1.5m0 0v-3.675A55 55 0 0 1 12 8.443m-7.007 11.55A5.98 5.98 0 0 0 6.75 15.75v-1.5"/></svg> Badge with Icon left  </span>"`;

exports[`Badge Component > renders Badge component with icon 3`] = `"<span class="sui-badge primary md outlined full">  Badge with Icon right <svg style="min-width: 16px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" height="16" width="16" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M4.26 10.147a60 60 0 0 0-.491 6.347A48.6 48.6 0 0 1 12 20.904a48.6 48.6 0 0 1 8.232-4.41a61 61 0 0 0-.491-6.347m-15.482 0a51 51 0 0 0-2.658-.813A60 60 0 0 1 12 3.493a60 60 0 0 1 10.399 5.84q-1.345.372-2.658.814m-15.482 0A51 51 0 0 1 12 13.489a50.7 50.7 0 0 1 7.74-3.342M6.75 15a.75.75 0 1 0 0-1.5a.75.75 0 0 0 0 1.5m0 0v-3.675A55 55 0 0 1 12 8.443m-7.007 11.55A5.98 5.98 0 0 0 6.75 15.75v-1.5"/></svg> </span>"`;
